predict.py

```diff
@@ -0,0 +1,187 @@
+            print(
+                f"WARNING! The output directory, {output_dir}, already exists, "
+                + "we will overwrite data in it!"
+            )
+        else:
+            raise FileExistsError(
+                f"The predictions directory, {output_dir}, already exists and isn't "
+                + "empty. We don't want to overwrite any existing results, exiting..."
+            )
+
+    for i, batch in enumerate(dataloader):
+        x = batch["image"].to(device)  # (N, B, C, H, W)
+        assert len(x.shape) in {4, 5}
+        if len(x.shape) == 5:
+            masks = []
+
+            def tensor_to_int(
+                tensor_tuple: Tuple[torch.Tensor, ...]
+            ) -> Tuple[int, ...]:
+                """Convert tuple of tensors to tuple of ints."""
+                return tuple(int(i.item()) for i in tensor_tuple)
+
+            original_shape = cast(
+                Tuple[int, int], tensor_to_int(batch["original_shape"])
+            )
+            patch_shape = cast(Tuple[int, int], tensor_to_int(batch["patch_shape"]))
+            padding = cast(Tuple[int, int], tensor_to_int(batch["padding"]))
+            patch_combine = CombineTensorPatches(
+                original_size=original_shape, window_size=patch_shape, unpadding=padding
+            )
+
+            for tile in x:
+                mask = task(tile)
+                mask = mask.argmax(dim=1)
+                masks.append(mask)
+
+            masks_arr = torch.stack(masks, dim=0)
+            masks_arr = masks_arr.unsqueeze(0)
+            masks_combined = patch_combine(masks_arr)[0]
+            filename = datamodule.predict_dataset.files[i]["image"]
+            write_mask(masks_combined, output_dir, filename)
+        else:
+            mask = task(x)
+            mask = mask.argmax(dim=1)
+            filename = datamodule.predict_dataset.files[i]["image"]
+            write_mask(mask, output_dir, filename)
+
+
+if __name__ == "__main__":
+    # Taken from https://github.com/pangeo-data/cog-best-practices
+    _rasterio_best_practices = {
+        "GDAL_DISABLE_READDIR_ON_OPEN": "EMPTY_DIR",
+        "AWS_NO_SIGN_REQUEST": "YES",
+        "GDAL_MAX_RAW_BLOCK_CACHE_SIZE": "200000000",
+        "GDAL_SWATH_SIZE": "200000000",
+        "VSI_CURL_CACHE_SIZE": "200000000",
+    }
+    os.environ.update(_rasterio_best_practices)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--config-dir",
+        type=str,
+        required=True,
+        help="Path to config-dir to load config and ckpt",
+    )
+
+    parser.add_argument(
+        "--predict_on",
+        type=str,
+        required=True,
+        help="Directory/Dataset to run inference on",
+    )
+
+    parser.add_argument(
+        "--output-dir",
+        type=str,
+        required=True,
+        help="Path to output_directory to save predicted mask geotiffs",
+    )
+
+    parser.add_argument("--device", type=str, default="cuda", choices=["cuda", "cpu"])
```

Comment:
Is this something we should let PyTorch Lightning handle?

Reply:
Perhaps. `LightningModules` shouldn't need this but I'll need to verify. Will get back to this.

Reply:
While `LightningModules` are aware of which device they're on, the models (UNet, DeepLab etc) are not. Since we can't forward device info this will be required.